% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/context_methylation_functions.r
\name{FilterContextCytosines}
\alias{FilterContextCytosines}
\title{Filter Cytosines in context}
\usage{
FilterContextCytosines(MethGR, genome, context)
}
\arguments{
\item{MethGR}{Granges obj of average methylation}

\item{genome}{BSgenome}

\item{context}{Context of interest (e.g. "GC", "CG",..)}
}
\value{
filtered Granges obj
}
\description{
Filter Cytosines in context
}
\examples{

Qinput = system.file("extdata", "QuasR_input_pairs.txt", package = "SingleMoleculeFootprinting", mustWork = T)
QuasRprj = GetQuasRprj(Qinput, BSgenome.Mmusculus.UCSC.mm10)
Samples = QuasR::alignments(QuasRprj)[[1]]$SampleName
sample = Samples[1]
MethGR = QuasR::qMeth(QuasRprj[grep(sample, Samples)], mode="allC", range, collapseBySample = TRUE, keepZero = TRUE)

FilterContextCytosines(MethGR, BSgenome.Mmusculus.UCSC.mm10, "NGCNN")

}
